Framework/API/src/DetectorSearcher.cpp

```diff
@@ -0,0 +1,224 @@
+  // choose the search strategy to use
+  if (!m_usingFullRayTrace) {
+    createDetectorCache();
+  } else {
+    m_rayTracer = Kernel::make_unique<InstrumentRayTracer>(instrument);
+  }
+}
+
+/** Create a NearestNeighbours search tree for the current instrument
+ */
+void DetectorSearcher::createDetectorCache() {
+  std::vector<Eigen::Vector3d> points;
+  points.reserve(m_detInfo.size());
+  m_indexMap.reserve(m_detInfo.size());
+
+  for (size_t pointNo = 0; pointNo < m_detInfo.size(); ++pointNo) {
+    if (m_detInfo.isMonitor(pointNo) || m_detInfo.isMasked(pointNo))
+      continue; // detector is a monitor or masked so don't use
+
+    // calculate a Q vector for each detector
+    // This follows a method similar to that used in IntegrateEllipsoids
+    const auto &det = m_detInfo.detector(pointNo);
+    const auto tt1 = det.getTwoTheta(V3D(0, 0, 0), V3D(0, 0, 1)); // two theta
+    const auto ph1 = det.getPhi();                                // phi
+    auto E1 =
+        V3D(-std::sin(tt1) * std::cos(ph1), -std::sin(tt1) * std::sin(ph1),
+            1. - std::cos(tt1)); // end of trajectory
+    E1 = E1 * (1. / E1.norm());  // normalize
+
+    Eigen::Vector3d point(E1[0], E1[1], E1[2]);
+
+    // Ignore nonsensical points
+    if (point.hasNaN())
+      continue;
+
+    points.push_back(point);
+    m_indexMap.push_back(pointNo);
+  }
+
+  // create KDtree of cached detector Q vectors
+  m_detectorCacheSearch =
+      Kernel::make_unique<Kernel::NearestNeighbours<3>>(points);
+}
+
+/** Find the index of a detector given a vector in Qlab space
+ *
+ * If no detector is found the first parameter of the returned tuple is false
+ *
+ * @param q :: the Qlab vector to find a detector for
+ * @return tuple with data <detector found, detector index>
+ */
+const DetectorSearcher::DetectorSearchResult
+DetectorSearcher::findDetectorIndex(const V3D &q) {
+  // quick check to see if this Q is valid
+  if (q.nullVector())
+    return std::make_tuple(false, 0);
+
+  // search using best strategy for current instrument
+  if (m_usingFullRayTrace) {
+    return searchUsingInstrumentRayTracing(q);
+  } else {
+    return searchUsingNearestNeighbours(q);
+  }
+}
+
+/** Find the index of a detector given a vector in Qlab space using a ray
+ * tracing search strategy
+ *
+ * If no detector is found the first parameter of the returned tuple is false
+ *
+ * @param q :: the Qlab vector to find a detector for
+ * @return tuple with data <detector found, detector index>
+ */
+const DetectorSearcher::DetectorSearchResult
+DetectorSearcher::searchUsingInstrumentRayTracing(const V3D &q) {
+  const auto direction = convertQtoDirection(q);
+  m_rayTracer->traceFromSample(direction);
+  const auto det = m_rayTracer->getDetectorResult();
+
+  if (!det)
+    return std::make_tuple(false, 0);
+
+  const auto detIndex = m_detInfo.indexOf(det->getID());
+
+  if (m_detInfo.isMasked(detIndex) || m_detInfo.isMonitor(detIndex))
+    return std::make_tuple(false, 0);
```

Comment:
This is a chicken & egg problem. We need to know which detector we hit before we know whether to discard it because the detector was masked/a monitor.